### ISSUE TITLE
Other Identifiers - add metadata

### ISSUE THREAD

**Issue by Jegelewicz** [ArctosDB/arctos#4101]
As I understand the proposal, currently we record other identifiers as follows:

**other_ID_type** | **other_ID_value** | other_ID_relationship 
 -- | -- | -- 
selected from code table | text added to base URL in the code table (when there is one) | relationship of this identifier to the cataloged item 

but if this is accepted, we would record other identifiers as follows:

other_ID_type | other_ID_text |  other_ID_value | other_ID_issued_by | other_ID_assigned_by | other_ID_assigned_date | other_ID_relationship 
 -- | -- | -- | -- | -- | -- | -- 
selected from code table | human readable ID | text added to base URL in the code table (when there is one) | Arctos Agent that issued the ID | Arctos Agent who assigned the ID to this record | date the identifier was assigned by the assigning agent | relationship inferred by this ID 

Under the mew method, only other_ID_type and other_ID_text would be required? Or should Aother_ID_assigned_by and other_ID_assigned_date also be required?

Am I missing something?

This will help us solve the problem of named identifiers that also have machine-readable IDs that have nothing to do with the name. See https://github.com/ArctosDB/arctos/issues/4055#issuecomment-956265421

> No. There will be no "value of the identifier" there will be - well, something, probably, likely more complex than "value," it needs worked out. Maybe 
> 
> * identifier_for_people (Santa Luzia)
> * identifier_for_machines (23166)
> * current stuff from the CT
> * [^^ that junk](https://github.com/ArctosDB/arctos/issues/4055#issuecomment-961292604)
> 
> and whatever else belongs in here, all as part of a single data object

_Originally posted by @dustymc in https://github.com/ArctosDB/arctos/issues/4055#issuecomment-961393222_

**Comment by Jegelewicz** [comment]
@campmlc @Nicole-Ridgwell-NMMNHS @dustymc please let me know if this is way off base!

**Comment by dustymc** [comment]
@jhpoelen knows of links from random stuff (inat) to Arctos - given sufficient metadata ("according to jorrit's scripts...") this could be inserted into Arctos via API.

Then: Use those links to grab data from inat (or whatever) to pull locality etc data for @ewommack

**Comment by campmlc** [comment]
And this could be a pub

**Comment by dustymc** [comment]
Current Model:

|column name|description|data type|null?|length|code table|
|-----------|-----------|---------|-----|------|----------|
|collection_object_id|FKEY cataloged_item|integer|NO|||
|other_id_number|integer component of identifier|numeric|YES|||
|other_id_type|Type of identifier|character varying|NO|75|[ctcoll_other_id_type](http://arctos.database.museum/info/ctDocumentation.cfm?table=ctcoll_other_id_type)|
|id_references|Relationship of the identifier to the attached record; establishes relationships between records|character varying|NO|60|[ctid_references](http://arctos.database.museum/info/ctDocumentation.cfm?table=ctid_references)|
|display_value|Assembled concatenation of (other_id_prefix,other_id_number,other_id_suffix)|character varying|NO|255||
|other_id_suffix|identifier suffix (after integer component)|character varying|YES|60||
|other_id_prefix|identifier prefix (before integer component)|character varying|YES|255||
|coll_obj_other_id_num_id|PKEY|integer|NO|||


Proposed model:


|column name|description|data type|null?|length|code table|
|-----------|-----------|---------|-----|------|----------|
|collection_object_id|FKEY cataloged_item|integer|NO|||
|other_id_number|numeric component of identifier|numeric|YES|||
|other_id_type|Type of identifier|character varying|NO|75|[ctcoll_other_id_type](http://arctos.database.museum/info/ctDocumentation.cfm?table=ctcoll_other_id_type)|
|id_references|Relationship of the identifier to the attached record; establishes relationships between records|character varying|NO|60|[ctid_references](http://arctos.database.museum/info/ctDocumentation.cfm?table=ctid_references)|
|display_value|Assembled concatenation of (other_id_prefix,other_id_number,other_id_suffix)|character varying|NO|255||
|other_id_suffix|identifier suffix (after integer component)|character varying|YES|60||
|other_id_prefix|identifier prefix (before integer component)|character varying|YES|255||
|coll_obj_other_id_num_id|PKEY|integer|NO|||
|preferred_display_value|User-supplied preferred (over display_value) display form|character varying|yes|255||
|issued_by_agent_id|FKEY-->agent; agent who issued the identifier|int|yes|||
|issued_on_date| date on which the identifier was issued|ISO8601|yes|||
|assigned_by_agent_id|FKEY-->agent; agent who assigned the identifier to the record|int|yes|||
|assigned_on_date|date on which the identifier was assigned|ISO8601|yes|||
|identifier_remarks|nonstructured information about the identifier which cannot be captured elsewhere|varchar|yes|||

> Or should Aother_ID_assigned_by and other_ID_assigned_date also be required?

I can see no reason to force the expansion on anyone, nor any reason to change the catalog record bulkloader, so I've proposed everything as NULLable. I think very few "normal" records will ever have any of these data, there's no reason to make most users supply some not-so-useful value, and the users who do need or want the additional information can use the ID bulkloader or directly edit the record.

One point of evaluation: Would the proposed information make collections comfortable with less-reliable (or "not reviewed by humans") identifiers assigned by agents like 'Arctos' (on behalf of my reciprocal relationship bot which currently emails bulkloadable files) or 'Globi' (some trusted but clearly 'not Arctos' application which could simply create these data - and so links to things like iNaturalist - via API)? If not, what could be added to accommodate that kind of usage?

I think "issued_by" (should be "minted by"???) is probably more complicated than the proposed model allows, but I don't know if that matters. MSB issues NKs to some department (external organization, project, whatever) who re-issues them to staff who re-issues them to students who assign them to dead rats: I think the only potentially-important bit of that is MSB, but if not now's the time to resolve.

See also https://github.com/ArctosDB/arctos/issues/2539#issuecomment-966816589 - I believe "biotic" is an attribute of the [authority](https://arctos.database.museum/info/ctDocumentation.cfm?table=ctid_references) and so DOES NOT need discussed here, but if that's not true, or not always true, then perhaps we need an additional concept (or determination - "AgentX thinks this [is|is not] biotic"), or something, in the data.

I remain entirely open to the idea that I've got all of this wrong, or that we need to revise the core/existing model in some fundamental way, or WHATEVER - anything which might influence how we model identifiers should be discussed here.

@campmlc this needs your careful attention, everything except preferred_display_value is from comments of yours which I may or may not have interpreted as intended.

**Comment by Jegelewicz** [comment]
> I can see no reason to force the expansion on anyone, nor any reason to change the catalog record bulkloader, so I've proposed everything as NULLable. 

Agree - but to make things complicated (or maybe not) should they be required when anything other than stuff in the current model is added? I guess I am thinking about this:

> Would the proposed information make collections comfortable with less-reliable (or "not reviewed by humans") identifiers assigned by agents like 'Arctos' (on behalf of my reciprocal relationship bot which currently emails bulkloadable files) or 'Globi' (some trusted but clearly 'not Arctos' application which could simply create these data - and so links to things like iNaturalist - via API)? If not, what could be added to accommodate that kind of usage?

Because as long as we know who (or what program) made the assertion, then we can adjust our confidence accordingly. Perhaps any IDs loaded with the "current model" are automatically assigned by the user who loaded them on the date they were loaded?

**Comment by Jegelewicz** [comment]
FWIW - the proposed model aligns with what I proposed to start off this issue and I think it all looks useful.

> I believe "biotic" is an attribute of the authority and so DOES NOT need discussed here, but if that's not true, or not always true, then perhaps we need an additional concept (or determination - "AgentX thinks this [is|is not] biotic"), or something, in the data.

I might think differently after a week or so, but "biotic" may be in the eye of the beholder? Some people might see "same lot as" as a biotic relationship and others may not? Same goes for "collected with". I sorta feel like we define our relationships and others can decided if they are "biotic" or not for their purposes?

**Comment by Jegelewicz** [comment]
Just saw this for "biotic" - http://www.ontobee.org/ontology/RO?iri=http://purl.obolibrary.org/obo/RO_0002574

**Comment by dustymc** [comment]
>know who (or what program) made the assertion

That would just be a condition of the API key (or similar) - we'd require them to assert as some agent, leave some remark, whatever.

>assigned by the user who loaded them

I think that would be very misleading - it just says "12" on some tag, the person behind the keyboard doesn't know more and probably won't have any more information in the future, they're just digitizing. (And "path to information or understanding" should be the entire point of the additional data.) If we want something like ``entered_by_agent_id`` then it should be added to the model. (I'm not sure I see the point but I don't think I'd have objections if someone else does.)

>define our relationships

That's kinda my take at the moment - X and Y should each be defined such that there's a clear boundary where the type is or is not appropriate, and from that "biotic enough" is left to the user. (I don't think that's the case now - a bunch of the 'same ... as' all look arbitrary and interchangeable to me.)

**Comment by campmlc** [comment]
Interesting

**Comment by Jegelewicz** [comment]
@dustymc something else just caught my eye. In #2539 John asked for use cases, so I just searched relationship "ate" and what I realize, is that (at least for this instance) there is absolutely no information about why this relationship was asserted.

https://arctos.database.museum/guid/CHAS:Herp:18616

Were the two things this animal supposedly "ate" found in it's stomach? Why should I have any confidence in the relationship? I think we should also think about relationship metadata and not rely on catalog record remarks to provide details. I don't know if this should be part of this issue or if it needs a new one?

**Comment by dustymc** [comment]
>absolutely no information 

``Collecting Method: stomach contents of CAS# 18616 (Sistrurus miliarius)``

>relationship metadata and not rely on catalog record remarks 

Yes, I think this is the correct place to discuss anything involving identifiers.

>Why should I have any confidence 

I don't see that changing much for this kind of use case; the method data (which seems correctly placed) could be replicated to the relationship metadata (where eventually one will be updated and then we'll never resolve the conflicting information....), but in this case I'm not sure how knowing more about agents and dates and such would add any confidence; everything involved has identifiers and such, and the numbers are "facts" (CHAS gave everything a primary identifier). Big-picture, I don't think extra metadata would be very useful most of the time (and the model I've proposed reflects that). I don't think there are any drawbacks to the metadata, and it's potentially very useful in fringe cases ("that looks like a 12 to me, but needs further scrutiny" instead of "12" become possible, scripts might be allowed more access, etc.) so I definitely support this idea, but I don't think its going to change anything fundamental about how we record what we find in a dead snake. (I'd be pretty happy to be proven wrong on that point!)

**Comment by Jegelewicz** [comment]
>     absolutely no information
> 
> Collecting Method: stomach contents of CAS# 18616 (Sistrurus miliarius)

Got it - but not sure these connections will get made because they end up in different places? Sometimes it is collecting method, sometimes collection object remark, sometimes in part remark! I don't want to conflate this with what we are doing here with identifiers, but I can see where it will eventually come up.

**Comment by jhpoelen** [comment]
@Jegelewicz @dustymc excited to hear all the discussions around the provenance (or origin) of claims in records. I am struggling to keep up, and I would like to think a little more on how to publish, cite, and retrieve the claims (and their provenance) independent of the record(s) that they happen to talk about. Perhaps a little live discussion with a bunch of napkins would be useful to come up with some practical ideas. Typing in little comment boxes in github issues is fun, but also has a bunch of limitations, especially when dealing with these innovative ideas. Up for a special session in the Arctos Den some time soon?

**Comment by campmlc** [comment]
Yes!

**Comment by dustymc** [comment]
>how to publish, cite, and retrieve the claims (and their provenance) independent of the record(s)

Doesn't seem problematic - identifiers (potentially including whatever we add via this discussion) are their own data objects, throwing up some API-or-whatever to treat them as such should be a pretty simple affair.

**Comment by jhpoelen** [comment]
I agree that figuring out a way to publish claims independently does/should not have to be problematic. 

How would you want GloBI to publish claims relating Arctos records?

**Comment by dustymc** [comment]
>how would you want GloBI to publish claims relating Arctos records?

This would set up the structure necessary to accept input (as opinions) from non-Curators, so barring any social issues (which will hopefully be worked out here as well) I'd like to see you add them directly to the Arctos catalog records. I intend to do the same with reciprocal relationships, which are now only suggestions (that may or may not ever get dealt with). That should be some relatively straightforward API (which could pull some data from the API key's metatdata - eg everything asserted through the API key you'd be issued magically gets asserted by Agent "GloBI"). Let me know if that doesn't seem realistic from some perspective.

**Comment by Jegelewicz** [comment]
> Perhaps a little live discussion with a bunch of napkins would be useful to come up with some practical ideas. Typing in little comment boxes in github issues is fun, but also has a bunch of limitations, especially when dealing with these innovative ideas. Up for a special session in the Arctos Den some time soon?

Soon-ish. This week is pretty full for me and I am visiting relatives next week for Thanksgiving. Week of November 28?

**Comment by Jegelewicz** [comment]
Maybe a monthly report of identifiers assigned by "magic" for follow up and ability to deny or confirm relationships suggested.

**Comment by Jegelewicz** [comment]
Added to AWG issues agenda 2021-12-02

**Comment by dustymc** [comment]
>Code Table Admins 

They're not who thought there was a problem with the old model?!?

>Added to AWG 

That is severely premature as well. I haven't heard much from the person who talks about this most, I've heard nothing from users on how the stuff mentioned in https://github.com/ArctosDB/arctos/issues/2539 might need incorporated, and I just don't think this is remotely in a place where it makes sense to talk about implementation.

>identifiers assigned by "magic" 

That's probably best addressed in https://github.com/ArctosDB/internal/issues/51.

**Comment by Jegelewicz** [comment]
AWG says GO but we also need a curatorial "agreement" flag.

**Comment by dustymc** [comment]
>curatorial "agreement" flag

Edited https://github.com/ArctosDB/arctos/issues/4101#issuecomment-967216485

I'd still like to have a careful review by @campmlc  (and anyone else who might use the expansion) before committing to this model.

**Comment by Jegelewicz** [comment]
Brought to the attention of Code Table peeps today.

**Comment by dustymc** [comment]
>Brought to the attention of Code Table peeps today.

The mention involved https://github.com/ArctosDB/arctos/issues/4101#issuecomment-969213595 - in short, if there's a place for "computers did this maybe don't trust it" then I could potentially magic in all sorts of derived relationships. @ebraker

**Comment by Jegelewicz** [comment]
We could lose a bunch of other identifier types if we used the "issued by" to associate "other identifiers" with the appropriate agent.

**Comment by dustymc** [comment]
I updated the [proposal](https://github.com/ArctosDB/arctos/issues/4101#issuecomment-967216485) - basically, make "accepted" more general.

>lose a bunch of other identifier types 

Here's hoping....

**Comment by Jegelewicz** [comment]
I like that!

I updated the [proposal](https://github.com/ArctosDB/arctos/issues/4101#issuecomment-967216485) - To add suggested labels for UI. @ccicero @campmlc please comment!

@dustymc any chance we could put this in place in test so that Carla and Mariel can see it in action?

**Comment by dustymc** [comment]
I added one more potential "me-field" to the proposal; I don't think it changes anything, just pays the CPU cost up front.

> put this in place in test

I don't think we're yet there; I'd still like that careful [readthrough](https://github.com/ArctosDB/arctos/issues/4101#issuecomment-967216485) by @campmlc, this mostly isn't my request and I'm not sure if I'm accurately conveying what I think I've been hearing or not.

I don't think the other concerns I laid out in the proposal comment have changed (or been addressed).

``assigned_by_agent_id``--->"agent who assigned the identifier to the Arctos record" needs more discussion;  that doesn't seem useful to me, I'd be more interested in who assigned the identifier to the whatever-got-cataloged. If we want a "pushed the button" Agent, that's probably best as another field (or pair of them, eg ``entered_by_agent_id fkey agent default session_user`` and ``entered_date DATE default current_date``).

**Comment by Jegelewicz** [comment]
@campmlc this one is waiting on a careful review by you.

**Comment by Jegelewicz** [comment]
Bring to AWG Issues

**Comment by Jegelewicz** [comment]
I need to develop a mock-up

**Comment by Jegelewicz** [comment]
Use https://arctos.database.museum/guid/ALMNH:Paleo:8993

**Comment by ccicero** [comment]
@dustymc Related to this, how can we make it easier to create relationships (apart from metadata)? I have ~300 bluebird blood samples, most with two or three-way relationships, e.g., same bird sampled three times requires 6 relationships. It's painful to do it manually. Can we somehow create magic, e.g., if they have the same identifier such as a band number, then magic the relationships?

Here's one example:

https://arctos.database.museum/guid/MVZ:Bird:186472
-- same individual as:
https://arctos.database.museum/guid/MVZ:Bird:186473
https://arctos.database.museum/guid/MVZ:Bird:186474

https://arctos.database.museum/guid/MVZ:Bird:186473
-- same individual as:
https://arctos.database.museum/guid/MVZ:Bird:186472
https://arctos.database.museum/guid/MVZ:Bird:186474

https://arctos.database.museum/guid/MVZ:Bird:186474
-- same individual as:
https://arctos.database.museum/guid/MVZ:Bird:186472
https://arctos.database.museum/guid/MVZ:Bird:186473

**Comment by dustymc** [comment]
>how can we ... create relationships

<img width="388" alt="Screen Shot 2022-03-21 at 5 19 01 PM" src="https://user-images.githubusercontent.com/5720791/159383591-b2cd08eb-8dee-44cb-b58f-45abe5453da1.png">

Basically, create one relationship, give the tool some time (a day I think) to find it, it will create data you can feed to the bulkloader.

> make it easier 

I could suggest more complex than simple reciprocals, but that would still require the user understanding what's being suggested, stripping out what doesn't make sense, using the tools, etc. I'd rather invest in the structure that can support full automation. Can we find a way to get this through? Maybe drop everything except the two extra columns I need (who and when) to do that, since there doesn't seem to be any non-theoretical interest in more?

**Comment by ccicero** [comment]
Hmm. OK, I'll give that a try, thanks.

So in the example I gave, just create one relationship between e.g., 
https://arctos.database.museum/guid/MVZ:Bird:186472
https://arctos.database.museum/guid/MVZ:Bird:186473

and it will find the relationship to https://arctos.database.museum/guid/MVZ:Bird:186474 and all the recripocals?

Or do I also need to create a relationship between 186472 and 186474 for the tool to work?

**Comment by dustymc** [comment]
You need to involve the third member. 

I will always need SOMETHING to get started. With this change I could go deeper - ``A parent of B`` and ``A parent of C`` could spawn ``B sibling of C`` (plus the reciprocal) instead of just the straight reciprocals, but I'll always need some assertion to start from. 

But, this could significantly broaden the scope of "some assertion" - eg I might be comfortable magicking 'same as' from BBL numbers (and then other stuff from that) if I had the metadata to undo them (and tell users where those data came from). If the data support that (and I have no idea if they do) then this could potentially save you a lot more work than just loading reciprocals.

And on that note, I think my bare minimum is actually three new fields - if I'm going to get that wild, I want remarks to explain it (in addition to who and when).

**Comment by campmlc** [comment]
Yes, agree we need remarks.

**Comment by jhpoelen** [comment]
hey y'all - I've been listening in on your conversations to make it easier to import relations from other sources. I'll rely on you to decide when would be a good time to prepare a GloBI test dataset to bulk import into Arctos. That'd be sooo cool to get that working and automated!

**Comment by campmlc** [comment]
Yes, that's the goal! And we'd have metadata to document the source.

**Comment by jhpoelen** [comment]
Can't wait to embed GloBI provenance tracking info in that documentation.

**Comment by campmlc** [comment]
Re @ccicero question, once there are more than three relationships, we can no longer manually create the links. Even three is too much. This is why we have the entity model, as a platform for displaying complex relationships. But having some kind of automation to form these through the link or search options in other collections would be hugely valuable.

**Comment by campmlc** [comment]
And yes, we certainly want GloBI and other provenance info clearly available!

**Comment by Jegelewicz** [comment]
> Maybe drop everything except the two extra columns I need (who and when) to do that, since there doesn't seem to be any non-theoretical interest in more?

YES plus a place for remarks. Can we just DO this part and get the GloBI test case running?

Current Model:

|column name|description|data type|null?|length|code table|
|-----------|-----------|---------|-----|------|----------|
|collection_object_id|FKEY cataloged_item|integer|NO|||
|other_id_number|integer component of identifier|numeric|YES|||
|other_id_type|Type of identifier|character varying|NO|75|[ctcoll_other_id_type](http://arctos.database.museum/info/ctDocumentation.cfm?table=ctcoll_other_id_type)|
|id_references|Relationship of the identifier to the attached record; establishes relationships between records|character varying|NO|60|[ctid_references](http://arctos.database.museum/info/ctDocumentation.cfm?table=ctid_references)|
|display_value|Assembled concatenation of (other_id_prefix,other_id_number,other_id_suffix)|character varying|NO|255||
|other_id_suffix|identifier suffix (after integer component)|character varying|YES|60||
|other_id_prefix|identifier prefix (before integer component)|character varying|YES|255||
|coll_obj_other_id_num_id|PKEY|integer|NO|||


Proposed model:


|column name|description|data type|null?|length|code table|UI label|
|-----------|-----------|---------|-----|------|----------|---|
|collection_object_id|FKEY cataloged_item|integer|NO|||N/A|
|other_id_number|numeric component of identifier|numeric|YES|||ID Number (int)|
|other_id_type|Type of identifier|character varying|NO|75|[ctcoll_other_id_type](http://arctos.database.museum/info/ctDocumentation.cfm?table=ctcoll_other_id_type)|ID Type|
|id_references|Relationship of the identifier to the attached record; establishes relationships between records|character varying|NO|60|[ctid_references](http://arctos.database.museum/info/ctDocumentation.cfm?table=ctid_references)|Relationship|
|display_value|Assembled concatenation of (other_id_prefix,other_id_number,other_id_suffix)|character varying|NO|255||N/A|
|other_id_suffix|identifier suffix (after integer component)|character varying|YES|60||Suffix|
|other_id_prefix|identifier prefix (before integer component)|character varying|YES|255||Prefix or string|
|coll_obj_other_id_num_id|PKEY|integer|NO|||N/A|
|**assigned_by_agent_id**|FKEY-->agent; agent who assigned the identifier to the **Arctos** record|int|yes|||ID Assigned By|
|**assigned_on_date**|date on which the identifier was assigned|ISO8601|yes|||ID Assigned Date|
|**Remarks**|

**Comment by dustymc** [comment]
>once there are more than three relationships, we can no longer manually create the links

That is not true, and I don't know why you think it is.

>entity model, as a platform for displaying complex relationships

Also not true.

> Can we just DO this part 

That's where I am - with those definitions - but I'm not going to want to immediately revisit the things we didn't do either so ???

**Comment by campmlc** [comment]
It may be technically possible to create more than three reciprocal
relationships using existing tools, but it is NOT logistically possible for
humans to do this, nor is it anything we could realistically expect Arctos
operators to tackle. Hence Carla's request.

**Comment by Jegelewicz** [comment]
UGH - I agree with @campmlc but I struggled to make anyone in @ArctosDB/arctos-code-table-administrators meeting "get" it and was tasked with "creating a mock-up". To do this, I need a good example of some GloBI thing that might use this - @jhpoelen  can you give me one?

The additional confusion came into play with how we would end up treating other identifier types, which I think we need to make as simple aspossible if we really want others to add this information to our records - Jorrit shouldn't have to wade through 500 identifiers to find the one that he thinks he should use to create the relationship.

**Comment by dustymc** [comment]
>NOT logistically possible

If you have an example I'd be happy to click the buttons. Arctos will handle most of the logistics. I'd still of course rather get this resolved to the point where I can just do it.

The holdup is [this mostly isn't my request and I'm not sure if I'm accurately conveying what I think I've been hearing or not](https://github.com/ArctosDB/arctos/issues/4101#issuecomment-1054723933).

We will certainly at some point need to work out what we call any new terms, but first we need to decide what we need to DO.

"Determiner" and "assigner" are different things, and "issuer," "reviewer,"  etc would be yet more. That pivots me back to my current position: Unless there is an engaged user with a clear-cut use case demanding some new concept, I vote for simplicity. @jhpoelen and I have the same requirements - assigner, assigned date, remarks, which are sufficient to un-do any messes we might make and explain the fuzzy bits to users. Everything else seems theoretical and murky and lacking use cases I say we drop it.


There was some confusion about how this would change agents (or something, I think I'm the confusee on that). It would not, although it could potentially provide more metadata - "this agent did xx identifier-things".

There was some confusion about how this would change how we handle resolvable identifiers. It would not.

This **could** change how we handle nonresolvable identifiers by allowing multiple agents to "issue" one type, instead of having hundreds of types. https://github.com/ArctosDB/arctos/issues/4450 strikes me as compelling evidence that The Community isn't ready for that, so I remain in favor of the minimal approach.

>Jorrit shouldn't have to wade through 500 identifiers

Sure he should, as should any user, assuming we need 500 identifiers. (We currently have 541 types, 305 are associated with URLs so don't really require documentation, the rest probably wouldn't be used by anyone DOING STUFF anyway - certainly not with the current documentation.) 

>creating a mock-up

Yea, I don't quite understand where that could go. This isn't much of a UI problem and can't be until we have a solid model, if nobody can show up with data that demands more than three additional fields then I'm still in favor of the minimum that would let us do stuff.

**Comment by campmlc** [comment]
The idea of reviewer etc comes from discussions that boil down to whether or not we can say an identifier and it's asserted relationship are "accepted" or not, and by whom on what date.  I do think that is valuable - as we have determined in other areas, such as events and taxonomy. This would be an extension of that concept to another set of shared data. 

During discussion of the ID issued by vs assigned by agents, we came back around to the idea of simplifying the other ID table and formalizing the relationship between IDs and agents. This has been proposed elsewhere as it is a recurring need. IF I remember our discussion correctly @Jegelewicz intended the ID of the issuing agent to be, for example "American Museum of Natural History", and the ID itself could be "AMNH 12345". Separating the ID itself from the institution or agent issuing it would drastically simplify ID table and make IDs more searchable.  We could have "institutional catalog number" or "collector number" and then have the issuing agent name linking to the agent table. The ID could be ZRC 12345 - and users would not necessarily have to know that is the identifier for "Lee Kong Chian Natural History Museum Zoological Reference Collection", as our current model requires. 

So, based on that, a simple use case would be the following? What else do we need? for GloBI? @jhpoelen 

ID Type: institutional catalog number
ID Value: MSB:Mamm:12345
ID url: https://arctos.database.museum/guid/MSB:Mamm:12345
ID issued by agent: Museum of Southwestern Biology Division of Mammals
ID Determined by agent: Jonathan L. Dunnum
ID Determined date: 2021-01-01
ID Accepted: yes
ID Accepted by agent: Mariel L. Campbell
ID Accepted date: 2022-03-22
ID remarks: formerly mistranscribed as MSB:Mamm:12453

**Comment by campmlc** [comment]
The other thing we have discussed with IDs is to make them linkable to events and to parts. Does that require another key?

**Comment by dustymc** [comment]
>say an identifier and it's asserted relationship are "accepted" or not, and by whom on what date.

So the question on that is how far do we want to go? Is one assertion (which can be flattened out into another column in the table) sufficient, or will we immediately want to expand that to "no, AgentB doesn't agree@date" and then "hold on here, AgentC says that's right@date" and so on and so forth?

If we can get by with the former then that's my preference - it's simple to model and simple to use - but I don't want to rebuild the very many UIs that use this and then immediately toss it all out either.

The more complicated/normalized model would provide a means to add more agent-stuff, which might be nice although I can't see any even sorta-realistic use cases. That would look something like

```
table id_agents
------------------------
pkey primary key
coll_obj_other_id_num_id references coll_obj_other_id_num(coll_obj_other_id_num_id)
agent_id references agent(agent_id)
agent_role references(some_new_code_table)
date is08601
```

and so adding a "agent-thing we didn't foresee" would be a matter of data (like adding a new attribute type) rather than structure.

>simplifying the other ID table

Elaborate? Everything here is more complications which we should tolerate only of they DO STUFF. (And I think they do, they let me add reciprocals and derivatives, provide a path for Globi-magic, etc.)

>The ID could be ZRC 12345 - and users would not necessarily have to know that is the identifier for "Lee Kong Chian Natural History Museum Zoological Reference Collection", as our current model requires.

At least in my view, identifiers are a way to get to other information, and that's it. Given "ZRC 12345" I can't get anywhere useful (is that a Nissan part number?) so I don't have much reason to care. Given Lee Kong Chian Natural History Museum Zoological Reference Collection ZRC 12345 I might be able to call them up, send them a letter, or eventually even link to their data. I don't think you can possibly be suggesting we break that link, but I can't figure out what you are suggesting.

> ID url: https://arctos.database.museum/guid/MSB:Mamm:12345

Are you suggesting we force users to type URLs? I think @Jegelewicz may be too, that sounds like a major unnecessary complication to me, I think I don't understand something. Minimally it would require more stuff in the catalog record bulkloader (at least if we have the "MSB:Mamm:12345" format somewhere - I'd be happy to see it go, https://github.com/ArctosDB/arctos/issues/4200); nothing I've proposed up to this point would require anything new from most users, it would just provide a place for it if you have something extra.

> linkable to events

Under the Entity model, just catalog the "critter-at-event" (aka Occurrence).

>and to parts.

I think part attributes does that sufficiently well, it would be some sort of a new structure if not. (Or https://github.com/ArctosDB/arctos/issues/3630 would provide a "special part identifier"). There's also a "what to catalog" component to this.

In either case, I think those two things are outside the scope of this conversation, which should be limited to identifiers which are keyed to catalog records (unless we want to toss out a big chunk of the model, which seems like fun but also more than we can handle at the moment).

**Comment by Jegelewicz** [comment]
Here is another way metadata might be useful. This record has two GenBank sequences, each from a different publication. Having each be assigned by lead author would help to demonstrate which came from which and the remark could hold a link to the publication?

![image](https://user-images.githubusercontent.com/5725767/160025052-71d39e05-729a-4db1-b1bd-4f7b756ecc27.png)

**Comment by campmlc** [comment]
Oh, great idea!

**Comment by dustymc** [comment]
> from this comment down

I think there are two big takeaways in there:

1. "fitting our collections into existing CMS" - I don't think there's anything mentioned that Arctos can't handle. Certainly some details would need fleshed out, but the core model is there and being used. I think we're anti-good at communicating that.
2. "representing [relationships] not as [simple stuff], but as [something actually capable of carrying the appropriate data]" - yup, we've (recently, surprisingly) figured that out too, this issue and all the bouncing around on Entities are good examples.

Speaking of, I think Entities are actually a more appropriate way of handling "big hunk of rock slowly being separated over time" than this Issue - there's not a lot of uncertainty (=potential metadata) in the relationship itself ("that Agent whacked it with a hammer and this thing fell off"), but the thing that fell off (and eventually the thing separated from it &etc.) needs catalog record level data that may or may not agree with the "ultimate source" (=Entity) data, along with (probably) part-attribute stuff (method, maybe-mostly). The distinction of course is in the metadata the objects are capable of carrying - these would plausibly (inevitably?) result in things like fine-scale lithostratigraphy determinations and citations, and that's the domain of catalog records.

Also "loanable objects" is probably a more intuitive term than my usual "thing to which one can stick barcodes" - I believe I'll steal it!

**Comment by Jegelewicz** [comment]
Check this out - http://bioguid.org/about

**Comment by campmlc** [comment]
Upping priority on this issue. Are we close to resolution?

**Comment by campmlc** [comment]
Maybe we need an "identifiers/relationships" module - not just a table - like our entity concept but just showing all identifiers and relationships with metadata and annotation tools.
BTW - GBIF is officially using the term "assertions" which have a place, time, agent.
In the meantime, can we just go with: https://github.com/ArctosDB/arctos/issues/4101#issuecomment-1075229290 ?

**Comment by campmlc** [comment]
https://github.com/ArctosDB/arctos/issues/4101#issuecomment-967216485

**Comment by dustymc** [comment]
>Are we close to resolution?

There are [at](https://github.com/ArctosDB/arctos/issues/4101#issuecomment-967216485) [least](https://github.com/ArctosDB/arctos/issues/4101#issuecomment-1075229290) [three](https://github.com/ArctosDB/arctos/issues/4101#issuecomment-1075544244) potential models in here, so no. There's also been some talk of requiring various proposed concepts, which would need a lot more discussion.

>can we just go with: https://github.com/ArctosDB/arctos/issues/4101#issuecomment-1075229290

That serves my purposes if that's the consensus, but this is going to be a LOT of work and we should really have something **solid** planned out before scrapping a big chunk of the core model; we're not going to want to immediately re-build whatever we build.

> "identifiers/relationships" module - not just a table

It's not "just a table" and hasn't been since we started linking to GenBank a couple decades ago. (IDs are visualized as a table in various places, but that has little to do with what the data do and can carry.)

**Comment by dustymc** [comment]
Well we've got at least three proposed models, why not a fourth? I propose a hybrid model.

Add to table  coll_obj_other_id_num two new columns:


|column name|default|description|data type|null?|UI label|
|-----------|------|---------|---------|----|---|
|**assigned_agent_id**|current_user|FKEY-->agent; agent who assigned the identifier to the **Arctos record**|int|yes|ID Assigned By|
|**assigned_date**|current_date|date on which the identifier was assigned to the **Arctos record**|date|yes|ID Assigned Date|

Second (from https://github.com/ArctosDB/arctos/issues/4101#issuecomment-1075544244), a new table

```
table identifier_agents_meta_thingee_table
------------------------
identifier_agent_id INT primary key
coll_obj_other_id_num_id REFERENCES coll_obj_other_id_num(coll_obj_other_id_num_id)
agent_id REFERENCES agent(agent_id)
agent_role REFERENCES some_new_code_table(some_new_column_name)
date iso8601
remark varchar
```

The coll_obj_other_id_num additions will meet the most core-ward "where did this come from?" use case, require no (or minor) changes to any UI or tool (we'll want to display it, but that's pretty trivial and doesn't have to be completed up front), and generally doesn't complicate any "normal" use cases. 

The new/extra table also requires no changes, just additions. New types of assertions are just code table entries and it's normalized - we can support that thing we haven't thought of yet or a decade of one agent saying "yes" and another saying "no," or 500 maybe-conflicting remarks (from 500 agent on 500 dates), or WHATEVER. This needs more consideration, I'm not sure what's proposed is correct at all - but that doesn't hold up the basic "who, when" core data that would let me magic identifiers even while we're working out the secondary details.

Bueno?

**Comment by Jegelewicz** [comment]
I get the first part and think it's a step in the right direction.

Is the second sort of an other identifier attribute?

**Comment by campmlc** [comment]
So the coll.obj other ID num ID references an agent, eg a collector or institution who provides the ID? So this is like the formerly suggested ID issued by? If so, I think I like it? If not, please explain?

**Comment by dustymc** [comment]
>econd sort of an other identifier attribute

Yup.

>references an agent

coll_obj_other_id_num_id references an identifier (the data object, not just values or whatever).
agent_id references an agent

>ike the formerly suggested ID issued by

That would be a role - a value in the new table

One identifier (object) might look like:

```
id: 1
type: MSB
value: 6
(eventually):assigned by you today
```
and then could have zero, 5 million, or anything in between attribute-like-agent-centric-thingees all referring to that one record....

```
references id: 1
agent: you
role: assigned by
date: whatever


references id: 1
agent: you
role: issued by
date: whatever


references id: 1
agent: you
role: whatever sort of value we might want in the new table
date: whatever


references id: 1
agent: you
role: some new thing that we add to the table in a few years
date: in a few years
```
 and display could be something like....

* MSB: 6
    *    assigned by you
    *    issued by you
    *    whatever sort of value we might want in the new table by you
    *    some new thing that we add to the table in a few years by you


Or the ID is a child of a catalog record, and the agent-involved-things are children of the ID.

I don't know if I'm making things more or less clear at this point, let's chat if the latter....

**Comment by Jegelewicz** [comment]
I think I get it seems like it would work.

**Comment by dustymc** [comment]
Entity meeting: GO! Entity-from-ids tool will need this to create complete data.

**Comment by dustymc** [comment]
I'm calling this next release.

All ways of creating identifiers should be capturing who and when.

Making those data visible on guid pages is is https://github.com/ArctosDB/arctos/issues/4605

Additional data is https://github.com/ArctosDB/arctos/issues/4604
